### PR TITLE
[resolver] use new resolver exceptions to create chain of responsibility

### PR DESCRIPTION
**Pull request by rotty3000**

### CODE REVIEW

File: biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java [hunk 1]
```diff
@@ -213,6 +215,106 @@ public static ResolutionException augment(ResolveContext context, ResolutionExce
...
+				// there will only be one entry here
+				chain.addAll(mre.getUnresolvedRequirements());
+			}
+
+			cause = cause.getCause();
+		}
+
+		Set<Requirement> mandatory = new HashSet<>();
+		Set<Requirement> optional = new HashSet<>();
+		for (Requirement req : re.getUnresolvedRequirements()) {
```
Comment (bjhargrave) [inline biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java]: You could use Collectors.partioningBy to partition the unresolved requirements.

File: biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java [hunk 2]
```diff
@@ -213,6 +215,106 @@ public static ResolutionException augment(ResolveContext context, ResolutionExce
...
+					}
+				});
+
+			if (!optional.isEmpty() && includeOptional) {
+				f.format("%nThe following requirements are optional:");
+				optional.stream()
+					.collect(Collectors.groupingBy(Requirement::getResource))
+					.forEach((r, l) -> {
+						f.format("%n    [%s]", r);
+						l.stream()
```
Comment (bjhargrave) [inline biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java]: You don't need a stream here. Just forEach on l.

File: biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java [hunk 3]
```diff
@@ -213,6 +215,106 @@ public static ResolutionException augment(ResolveContext context, ResolutionExce
...
+	}
+
+	/**
+	 * This output doesn't produce an appropriate 'chain of responsibility' for
+	 * resolution failures. See {@link #format(ResolutionException)}.
+	 *
+	 * @param requirements the unresolved requirements
+	 * @return the report
+	 */
+	@Deprecated
```
Comment (bjhargrave) [inline biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java]: If it is deprecated, why alter its behavior?

File: maven/bnd-resolver-maven-plugin/README.md [hunk 4]
```diff
@@ -58,3 +58,4 @@ mvn bnd-resolver:resolve
 |`failOnChanges`        | Whether to fail the build if any change in the resolved `-runbundles` is discovered. _Defaults to `true`._|
 |`bundles`              | This is the collection of files to use for locating bundles during the bndrun resolution. Paths are relative to `${project.basedir}` by default. Absolute paths are allowed. _Defaults to dependencies in the `compile` and `runtime`, plus the current artifact (if any)._|
 |`useMavenDependencies` | If `true`, adds the project's compile and runtime dependencies to the collection of files to use for locating bundles during the bndrun resolution. _Defaults to `true`._|
+|`includeOptional`      | If `true`, failure reports will include optional requirements. _Defaults to `true`._|
```
Comment (bjhargrave) [inline maven/bnd-resolver-maven-plugin/README.md]: Same comment as for gradle. `reportOptional` is a better name.

File: biz.aQute.bnd.gradle/README.md [hunk 5]
```diff
@@ -410,6 +410,11 @@ The collection of files to use for locating bundles during the
 bndrun resolution. The default is _${project.sourceSets.main.runtimeClasspath}_
 plus _${project.configurations.archives.artifacts.files}_.
 
+### includeOptional
```
Comment (bjhargrave) [inline biz.aQute.bnd.gradle/README.md]: `reportOptional` is a better name.

File: biz.aQute.bnd.gradle/src/aQute/bnd/gradle/Resolve.groovy [hunk 6]
```diff
@@ -60,6 +62,7 @@ public class Resolve extends DefaultTask {
   boolean failOnChanges
 
   private File bndrun
+  private boolean includeOptional = true
```
[the remark this answers is not in the document]
Comment (bjhargrave) [reply biz.aQute.bnd.gradle/src/aQute/bnd/gradle/Resolve.groovy]: What is the purpose of this option? Just to make the output smaller? At first I thought this had to do with ResolveContext.getOptionalResources but it is just about the resolution directive value.

**Comment by rotty3000** [comment]
@bjhargrave updated

**Pull request merged**